### PR TITLE
Implement MSC3818: copy room type on upgrade

### PR DESCRIPTION
**Pull request by Mikaela**
Resolves: #11896

  Signed-off-by: Aminda Suomalainen

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). 
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -469,13 +469,17 @@ async def clone_existing_room(
             (EventTypes.PowerLevels, ""),
         ]
 
-        # If the old room was a space, copy over the room type and the rooms in
-        # the space.
+        # Copy the room type as per MSC3818 https://github.com/matrix-org/matrix-spec-proposals/pull/3818.
```
Comment (squahtx) [inline synapse/handlers/room.py:472]: style nit: Let's avoid overly long lines.
```suggestion
        # Copy the room type as per MSC3818
        # https://github.com/matrix-org/matrix-spec-proposals/pull/3818.
```

File: changelog.d/12786.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Implement [MSC 3818: Copy room type on upgrade](https://github.com/matrix-org/matrix-spec-proposals/pull/3818).
```
Comment (squahtx) [inline changelog.d/12786.feature:1]: style nit: We usually don't insert a space in MSC names.
```suggestion
Implement [MSC3818: Copy room type on upgrade](https://github.com/matrix-org/matrix-spec-proposals/pull/3818).
```

**Review by squahtx — APPROVED** [review]
Looks good! Just two minor suggestions.

I think this is good to merge before the MSC progresses, as the current spec leaves the room upgrade behaviour up to implementations, according to https://github.com/matrix-org/synapse/issues/10207#issuecomment-914505834

File: synapse/handlers/room.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/room.py:472]: I think we can just not include the URL -- we usually don't in our code comments.

File: synapse/handlers/room.py [hunk 3]
```diff
@@ -469,13 +469,17 @@ async def clone_existing_room(
...
+        if old_room_create_event.content.get(EventContentFields.ROOM_TYPE) is not None:
```
Comment (clokep) [inline synapse/handlers/room.py:473]: Can we pull the field out once and re-use it in the two if-statements please.

```python
room_type = old_room_create_event.content.get(EventContentFields.ROOM_TYPE)
if room_type is not None:
    ...
if room_type == RoomTypes.SPACE:
    ...
```

Comment (clokep) [reply synapse/handlers/room.py:473]: (Can likely indent the second if-statement too, but that's not super important!)

File: synapse/handlers/room.py (hunk 1 above)
Comment (turt2live) [reply synapse/handlers/room.py:472]: (how do you survive!?!?!?!? I'd go insane if I couldn't click links directly to MSCs)

**Comment by Mikaela** [comment]
I think I resolved the expressed concerns on this side

**Comment by squahtx** [comment]
Looks good to merge. Thanks for contributing to Synapse!

**Pull request merged**